aiohttp/client.py

```diff
@@ -415,8 +415,8 @@ def _build_url(self, str_or_url: StrOrURL) -> URL:
         if self._base_url is None:
             return url
         else:
-            assert not url.absolute and url.path.startswith("/")
-            return self._base_url.join(url)
+            assert not url.absolute
+            return self._base_url.join(URL(url.path_qs.lstrip("/")))
```

Comment:
Hmm, I guess this is another complication. The spec suggests that `/foo` should replace the base path, while `foo` would get merged onto the base path.

@bdraco How do you think it should work here?

Reply:
Reconstructing the URL again here is going to have an impact on every request.   It seems whats wanted it a new function in yarl to `append` urls instead of `join` them.


Side note .. URL processing used to be ~15-25% of the request cycle in some cases, and we have done a lot to improve the performance there recently.

Reply:
> It seems whats wanted it a new function in yarl to `append` urls instead of `join` them.

Well, my first question is whether that should be the behaviour? Or should we stick with the RFC behaviour of merging (i.e. the same behaviour in browsers)? In which case this change can be reverted.

Reply:
It looks like the docs don't say how the URL will be joined to the `base_url`
https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession

It gets messy quickly once we start modifying the URL with string processing. For example, the current implementation in this PR is dropping the fragment because `path_qs` does not return it.

I think we shouldn't alter the url path and joining should be fully governed by the semantics of `URL.join`

Reply:
> It looks like the docs don't say how the URL will be joined to the `base_url` https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession

We explicitly disallowed paths previously, likely due to these concerns. So, this is the point that we decide how to do it.

> I think we shouldn't alter the url path and joining should be fully governed by the semantics of `URL.join`

Right, so docs will need to explain this, and the most common usecase will be paths without a leading slash then.